include/seqan3/io/sequence/sequence_file_in.hpp

```diff
@@ -1,53 +1,113 @@
+// ==========================================================================
+//                 SeqAn - The Library for Sequence Analysis
+// ==========================================================================
+//
+// Copyright (c) 2006-2017, Knut Reinert, FU Berlin
+// Copyright (c) 2016-2017, Knut Reinert & MPI Molekulare Genetik
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+// Author: Svenja, Temesgen, Jongkyu <our_emails@fu-berlin.de>
+// ==========================================================================
+// sequence_file_in.hpp
+// ==========================================================================
+
 #pragma once
 
+#include <fstream>
 #include <string>
 #include <variant>
 #include <vector>
+#include <functional>
+#include "sequence_file_format.hpp"
+#include "sequence_file_format_fasta.hpp"
 
 namespace seqan3
 {
 
 // ==================================================================
 // sequence_file_in_traits
 // ==================================================================
-
 template <typename t>
 concept bool sequence_file_in_traits_concept = requires (t v)
 {
-    t::stream_type;
-    t::valid_formats;
+    typename t::stream_type;
+    typename t::valid_formats;
+    // //TODO valid_formats shall be variant and all of variants types
```

Comment:
you need to add a meta-function like ```is_valid_format_type``` which can check that the type is a variant and that can require the contained types to fulfil the format_concept.